### PR TITLE
[mlir][emitc] Refactor ArithToEmitC: perform sign adaptation, type conversions / cast insertion in a single place

### PR DESCRIPTION
**Pull request by cferry-AMD**
Factor EmitC type signedness adaptation and cast operations in ArithToEmitC using adaptValueType and adaptIntegralTypeSignedness.

### CODE REVIEW

File: mlir/include/mlir/Dialect/EmitC/IR/EmitC.td [hunk 1]
```diff
@@ -288,6 +288,7 @@ def EmitC_CastOp : EmitC_Op<"cast",
   let arguments = (ins EmitCType:$source);
   let results = (outs EmitCType:$dest);
   let assemblyFormat = "$source attr-dict `:` type($source) `to` type($dest)";
+  let hasFolder = 1;
```
Comment (simon-camp) [inline mlir/include/mlir/Dialect/EmitC/IR/EmitC.td:291]: Why is this needed? I would have thought folding is done through the CastOpInterface, but I haven't checked.

Comment (cferry-AMD) [reply mlir/include/mlir/Dialect/EmitC/IR/EmitC.td:291]: Good catch -- looks like this folder works. I don't know why it didn't happen during my tests...

**Review by simon-camp — APPROVED** [review]

**Comment by cferry-AMD** [comment]
Thanks @simon-camp. Let's merge this refactor with a single review, we'll need to spend more time discussing #95795 so further updates can land in there as well.
